File: src/test/java/org/meetkt/invoice/domain/model/InvoiceUnitTest.java
package org.meetkt.invoice.domain.model;

import org.junit.jupiter.api.Test;
import org.meetkt.basket.domain.model.Basket;
import org.meetkt.catalogue.domain.model.ProductId;
import org.meetkt.catalogue.fixture.ProductFixture;
import org.meetkt.basket.domain.model.Price;

import static org.assertj.core.api.Assertions.assertThat;

class InvoiceUnitTest {

    @Test
    void shouldReturnTotalItems() {
        Basket basket = Basket.empty();
        basket.add(ProductFixture.aProduct().withProductId("P-001").withPriceInt(10).build());
        basket.add(ProductFixture.aProduct().withProductId("P-002").withPriceDouble(20.78).build());

        Invoice invoice = Invoice.of(basket);
        int totalItems = invoice.totalItems();

        assertThat(totalItems).isEqualTo(2);
    }

    @Test
    void shouldReturnTotalPrice() {
        Basket basket = Basket.empty();
        basket.add(ProductFixture.aProduct().withProductId("P-001").withPriceDouble(20.78).build());
        basket.add(ProductFixture.aProduct().withProductId("P-001").withPriceDouble(20.78).build());

        Invoice invoice = Invoice.of(basket);
        Price totalPrice = invoice.totalPrice();

        assertThat(totalPrice).isEqualTo(new Price(41.56));
    }

    @Test
    void shouldReturnTotalItemsForAProductId() {
        Basket basket = Basket.empty();
        basket.add(ProductFixture.aProduct().withProductId("P-001").withPriceDouble(20.78).build());
        basket.add(ProductFixture.aProduct().withProductId("P-001").withPriceDouble(20.78).build());

        Invoice invoice = Invoice.of(basket);
        int totalItems = invoice.totalItemsFor(new ProductId("P-001"));

        assertThat(totalItems).isEqualTo(2);
    }
}

File: src/test/kotlin/in/meetkt/invoice/domain/model/InvoiceUnitTest.kt
package `in`.meetkt.invoice.domain.model

import `in`.meetkt.basket.domain.model.Basket
import `in`.meetkt.basket.domain.model.Price
import `in`.meetkt.catalogue.domain.model.ProductId
import `in`.meetkt.catalogue.fixture.ProductFixture
import org.assertj.core.api.Assertions.assertThat
import org.junit.jupiter.api.Test

class InvoiceUnitTest {

    @Test
    fun shouldReturnTotalItems() {
        val basket = Basket.empty()
        basket.add(ProductFixture.aProduct().withProductId("001").withPriceInt(10).build())
        basket.add(ProductFixture.aProduct().withProductId("002").withPriceDouble(20.78).build())

        val invoice = Invoice.`for`(basket)

        val totalItems = invoice.totalItems()
        assertThat(totalItems).isEqualTo(2)
    }

    @Test
    fun shouldReturnTotalPrice() {
        val basket = Basket.empty()
        basket.add(ProductFixture.aProduct().withProductId("001").withPriceDouble(20.78).build())
        basket.add(ProductFixture.aProduct().withProductId("001").withPriceDouble(20.78).build())

        val invoice = Invoice.`for`(basket)

        val totalPrice = invoice.totalPrice()
        assertThat(totalPrice).isEqualTo(Price(41.56))
    }

    @Test
    fun shouldReturnTotalItemsForAProductId() {
        val basket = Basket.empty()
        basket.add(ProductFixture.aProduct().withProductId("001").withPriceDouble(20.78).build())
        basket.add(ProductFixture.aProduct().withProductId("001").withPriceDouble(20.78).build())

        val invoice = Invoice.`for`(basket)
        val totalItems = invoice.totalItemsFor(ProductId("001"))

        assertThat(totalItems).isEqualTo(2)
    }
}